### ISSUE TITLE
Detect unnecessary function return type declaration

### ISSUE THREAD

**Issue by JKillian** [palantir/tslint#695]
The `no-inferrable-types` rule warns when variable or parameter declarations have a declared type that is directly inferable. A similar feature could be added for function return types.

For example, if a function doesn't return, its return type can be inferred as void and an explicit `: void` declaration in unnecessary.

**Comment by adidahiya** [comment]
needs some elaboration on how the control flow analysis would work, whether we would share code with `no-unreachable`, etc.

**Comment by myitcv** [comment]
Probably need to implement this using the `TypeChecker` to ask it what the return type of the function is. If the return type is `void`, then you can check that the annotation is not present on the function declaration.

**Comment by ibezkrovnyi** [comment]
Implemented rule `no-inferrable-return-types` ([github](https://github.com/ibezkrovnyi/tslint-rules)/[npm](https://www.npmjs.com/package/@ibezkrovnyi/tslint-rules)).

It is my first tslint/typescript related project. I didn't found possibility to infer return type using `TypeChecker`, so I tricked `getReturnTypeOfSignature`. Simple `typesAreEqual` from https://github.com/ajafff/tslint-consistent-codestyle (@ajafff) is used.

What can be improved:
* TS compiler already have `checkTypeRelatedTo`, `isRelatedTo`, `isIdenticalTo` internally
* TS compiler has code inside `getReturnTypeOfSignature`, which infers return type

It would be much better if we had them exported.

However it seems like rule already works for my cases, tested with TypeScript 2.7.x.
